Update realm icon

FadhlanR opened the pull request:
This PR is made to update the styling of the realm icon in the stack item's header of operator mode when the stack item is on edit view. But I also refactored the way we load the realm info by utilizing `RealmInfoProvider`.

Screenshots:

- In stack item of operator mode when edit view.

<img width="242" alt="Screenshot 2023-10-02 at 19 14 55" src="https://github.com/cardstack/boxel/assets/12637010/d5a7ebf0-e28d-4e6b-857c-19e8a0e1c395">

- In stack item of operator mode when isolated view.

<img width="256" alt="Screenshot 2023-10-02 at 19 15 01" src="https://github.com/cardstack/boxel/assets/12637010/602fa6e6-0ead-45a5-863f-81dd10a3e52b">

- in card URL Bar.

<img width="183" alt="Screenshot 2023-10-02 at 19 15 06" src="https://github.com/cardstack/boxel/assets/12637010/ba53e5d6-cf6c-4394-9b93-59b9e142e964">

FadhlanR committed:
Update realm icon

FadhlanR committed:
test fixes

FadhlanR committed:
remove unnecessary div

habdelra commented on packages/host/app/components/operator-mode/stack-item.gts:
```diff
@@ -288,147 +267,154 @@ export default class OperatorModeStackItem extends Component<Signature> {
-            <div>
-              <BoxelDropdown>
-                <:trigger as |bindings|>
+          <:ready as |realmInfo|>
```
what do we show when the `realmInfo` is being determined (since it's async)?

FadhlanR replied:
Hmm, we don't display anything in the header. Do you think it would be better to revert to the old approach rather than using the `RealmInfoProvider` for this case?

habdelra replied:
the old way was synchronous, yes? if so then yeah it's probably a better idea

FadhlanR replied:
No, it's asynchronous as well https://github.com/cardstack/boxel/pull/700/files#diff-ca60a92fb9d0b21d7b89b8245f7a84b3078b4585c6b47ea474a536d7bcedd984L183. However, in the old way, we could still display the entire header, although without the right realm icon and realm name.

FadhlanR replied:
Updated [f2df233](https://github.com/cardstack/boxel/pull/700/commits/f2df233217957ff8dec66b910b976725c8cc4eff)

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -38,11 +37,17 @@ export default class CardURLBar extends Component<Signature> {
-          {{if this.realmName this.realmName 'Unknown Workspace'}}</span>
+        <RealmInfoProvider
+          @realmURL={{this.realmURL}}
+          @returnUnknownRealmIfError={{true}}
```
You can use the `error` block in `RealmInfoProvider` to provide custom error displays.

I think displaying "Unknown workspace" could be confusing to the user. If there is an error while fetching the realm isn't it better to say exactly that?

jurgenwerk reviewed (commented):
Sorry, approved by mistake. I have some confusion regarding your approach of catching the error while fetching the realm info and making it look fine (yielding to the `ready` block) to the consumer of the realm info provider. Isn't it more suitable to use the error block for that?

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -194,11 +200,14 @@ export default class CardURLBar extends Component<Signature> {
+      url = this.cardService.getRealmURLFor(new URL(this.codePath))?.toString();
+    }
+    if (!url) {
+      url = this.cardService.defaultURL.toString();
```
When the realm could not be resolved from the given file URL, is it really correct to fall back to the default realm url?

FadhlanR replied:
I made it based on an existing test case.

jurgenwerk replied:
If I understand correctly, in case the realm will fail to be resolved for the given path, you want to return the default realm url? How come?

FadhlanR replied:
Yes based on this test case https://github.com/cardstack/boxel/blob/6108d83241f5c8b91f190616c24d7a9b3f822599/packages/host/tests/acceptance/code-mode-test.ts#L839

jurgenwerk replied:
Hmm, it seems that I mistakenly thought you are using `fetchRealmURL` from the `RealmInfoService` - so I was worried that if that fetch fails for some reason (network error?) then some other realm would be used, possibly wrong one (if the default one differs from the one that the code path is in)

FadhlanR replied:
I see, No, I'm not using `fetchRealmURL` from `RealmInfoService`. I decided to use `realmURL` from `CodeMode` component to make it consistent. [2864307](https://github.com/cardstack/boxel/pull/700/commits/28643078f4dee6ae6f04f3f77dff2e951b3905c8)

jurgenwerk replied:
Will figuring out the realm url work when there is a codePath that belongs to a realm which we do not have stored in `ownRealmURL` or `otherRealmURLs`?

FadhlanR replied:
I think so, because `realmURL` in `CodeMode` component is from a file that is opened. https://github.com/cardstack/boxel/blob/6108d83241f5c8b91f190616c24d7a9b3f822599/packages/host/app/components/operator-mode/code-mode.gts#L234

FadhlanR committed:
Revert unnecessary changes

FadhlanR commented:
> Sorry, approved by mistake. I have some confusion regarding your approach of catching the error while fetching the realm info and making it look fine (yielding to the `ready` block) to the consumer of the realm info provider. Isn't it more suitable to use the error block for that?

I reverted those changes here [f2df233](https://github.com/cardstack/boxel/pull/700/commits/f2df233217957ff8dec66b910b976725c8cc4eff). I added `returnUnknownRealmIfError ` because I want to make "Unknown Workspace" value to be general so I could use it on `card-url-bar` and `stack-item`. However, I decided not to implement `RealmInfoProvider` in `stack-item`, so I don't need `returnUnknownRealmIfError` anymore.

FadhlanR committed:
Merge branch 'main' into cs-5955-update-realm-icon

FadhlanR committed:
Use realm URL from code-mode

jurgenwerk commented on packages/host/app/components/operator-mode/card-url-bar.gts:
```diff
@@ -38,11 +38,20 @@ export default class CardURLBar extends Component<Signature> {
+          </:ready>
+          <:error>
+            <div class='realm-icon'>
+              <img src='' alt='realm-icon' />
```
Is img with a blank src valid html?

FadhlanR replied:
Replaced it [254f2f2](https://github.com/cardstack/boxel/pull/700/commits/254f2f2c9a8b1ea6f0b6e9c2d900a42766f1ce2e). It will look like this.

![Screenshot 2023-10-03 at 19 26 43](https://github.com/cardstack/boxel/assets/12637010/b88f78a1-7c03-4584-9099-7a66edf3fa53)

FadhlanR committed:
Update unknow workspace icon
